feat(apikey): implement API key table, object model, and service layer

wenyang-cao opened the pull request:
This pull request implements the following features related to API key management:

Create APIKEY table and handle schema migration.

Add API keys to scala object model.

Add service layer methods for interfacing object model with database operations - are going to be needed following methods:

creating API key for given user
deleting API key by id
fetching API key by id
fetching API keys for given user
fetching API keys for given organization
lookup API key based on API key value

Tests:
Functional Tests: The new API endpoints were manually tested using Postman..

Unit Tests: Unit tests were added to cover key methods within the service layer, ensuring that the logic for handling API keys is functioning correctly.
![image](https://github.com/user-attachments/assets/ee9bd57f-365e-4b04-9237-e55eb76e6718)

To Do:
Authentication Support for API Keys: Modify the authentication module to support API key-based authentication for the entire API. 

For further details please refer to the design doc.

kroczi reviewed (commented):
@wenyang-cao Very nice job. I have added a few comments.

wenyang-cao commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/OrgApiKeys.scala:
```diff
@@ -0,0 +1,83 @@
+  )
+  val orgApiKeys: Route =
+  path("orgs" / Segment / "apikeys") { orgid =>
+    onSuccess(db.run(OrgsTQ.getOrgid(orgid).result)) {
+      case Nil =>
+        complete(HttpCode.NOT_FOUND, ApiResponse(ApiRespType.NOT_FOUND, ExchMsg.translate("org.not.found", orgid)))
+      case _ =>
+        exchAuth(TOrg(orgid), Access.READ) { identity =>
+          if (identity.getOrg == orgid && identity.isAdmin)
+            getOrgApiKeys(identity, orgid)
+          else
+            complete(HttpCode.ACCESS_DENIED, ApiResponse(ApiRespType.ACCESS_DENIED, ExchMsg.translate("access.denied.not.org.admin")))
+        }
+    }
```
The current permission approach for this API is different from the one used for the GET /orgs/{organization}/users endpoint. The user endpoint allows access to be granted to the root user, org admins, and hub admins. However, for the apikeys endpoint,in the Q&A the question explicitly asks whether a hub admin should have cross-org privileges, and the answer states that hub admins do not have such privileges and should be treated as regular users when accessing resources outside their own organization. This distinction means that exchAuth(TUser(OrgAndId(orgid, "#").toString), Access.READ) will not work as expected, as it doesn't enforce the restriction for hub admins.

wenyang-cao replied:
To be honest, I couldn't find a specific permission that restricts access to only the admin of a particular organization without allowing roles like Hub Admin or SuperUser to access it. In order to remove the explicit call to identity.isAdmin and rely entirely on the permissions mechanism, I currently modify the code and used exchAuth(TOrg(orgid), Access.WRITE), even though its name is not appropriate. This permission does effectively restrict access to the organization admin as required, and all unit tests passed successfully. Although this works as expected, I haven’t found a perfectly fitting permission for this case yet.

If there is a more appropriate permission available, please let me know,  and I'll update the implementation accordingly. I tested a few other options that seemed suitable by name, but they didn't behave as expected.

Alternatively, if no such permission exists, we may consider extending the permissions enumeration to better handle access control for reading API keys in an organization in the future.

kroczi replied:
By the suggested change it will result in `Access.READ_MY_USERS` access required:
```
  override def authorizeTo(target: Target, access: Access)(implicit logger: LoggingAdapter): Try[Identity] = {
    val requiredAccess: Access =
      // Transform any generic access into specific access
      if (isMyOrg(target) || target.isPublic) {
        target match {
          case TUser(id) => access match { // a user accessing a user
            case Access.READ => logger.debug(s"authorizeTo(): target id=$id, identity creds.id=${creds.id}")
              if (id == creds.id) Access.READ_MYSELF
              // since we are in the section in which identity and target are in the same org, if identity is a hub admin we are viewing the users in the root org. Note: the root user is also an hub admin and org admin.
              else if (isHubAdmin || target.mine) Access.READ_MY_USERS // the routes's getAnyProblem() methods will catch the rest of the hubAdmin cases
              else Access.READ_ALL_USERS
       ...
      } else if (isHubAdmin) { // cross-org access is "normal" for a hub admin, because the hub admin is defined in the root org
        // since we are in the cross-org section, the target will never be itself or root
        target match {
          case TUser(id) => access match { // a hub admin accessing a user
              case Access.READ => logger.debug(s"authorizeTo(): target id=$id, identity creds.id=${creds.id}")
                Access.READ_MY_USERS // the get routes filter out regular users
        ....
```

This will allow `AdminUser` from that org who has `ALL_IN_ORG` access, `HubAdmin` (regardless of their org) who has `READ_MY_USERS` as well as `SuperUser` who has `ALL`. But will reject regular `User` who has only `READ_MYSELF` access and AdminUsers from other orgs.

I believe this will be fine. The design doc is not precise in this regard in this area, I need to fix that. @naphelps could you please comment on this?

kroczi replied:
@naphelps What are your thoughts on the permissions?

naphelps replied:
Check the comment I made today above. Should probably keep the same as the top-most user route.

wenyang-cao commented on src/main/scala/org/openhorizon/exchangeapi/table/apikey/ApiKeys.scala:
```diff
@@ -0,0 +1,17 @@
+class ApiKeys(tag: Tag) extends Table[ApiKeyRow](tag, "apikeys") {
+  def orgid = column[String]("orgid")
+  def id = column[String]("id", O.PrimaryKey) // UUID v4 string
+  def username = column[String]("username")   // composite id: org/user
+  def description = column[String]("description")
```
Just to clarify, for username only or for both fields? Currently, we don't have any specific queries that involve the 'description' field

wenyang-cao commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -0,0 +1,145 @@
+  val userApiKeys: Route = pathPrefix("orgs" / Segment / "users" / Segment / "apikeys") { (orgid, username) =>
+    pathEndOrSingleSlash {
+      post {                                 //Should create enumeration later
+        exchAuth(TUser(s"$orgid/$username"), Access.WRITE) { identity =>
+        if (!(identity.creds.id == s"$orgid/$username" || (identity.getOrg == orgid && identity.isAdmin)))
+            complete(HttpCode.ACCESS_DENIED, ApiResponse(ApiRespType.ACCESS_DENIED, "Only the user or org admin can create API keys."))
+          else
+            postUserApiKey(identity, orgid, username)
+        }
```
I think only rely on permissions should work. Initially I added those check just to ensure everything as expected since I was not so familiar with those permissions enumerations and I was not sure would it behave as expected. I have tried to remove those explicit checks the tests still pass.

wenyang-cao commented:
Update Summary
This new commit includes the following changes and improvements:

Completed and verified Swagger annotations for all new and existing API key–related endpoints. Ensured they are properly rendered in the Swagger UI with correct request/response examples and schemas

Added indexes to the apikeys table

Refactored authorization logic by removing explicit identity.isAdmin checks. Instead, the access control is now consistently handled through the exchAuth(...) wrapper, aligning with the project-wide permission model.

All unit tests still pass successfully.

To do and to further discuss:
1.Should addition/removal of API key touch the user (change its lastUpdated timestamp?)  
This commit does not implement this behavior, as the decision is still under consideration. My current stance is that updating lastUpdated is not necessary, since API keys are stored in a separate table (apikeys) and are logically independent from the users table — they do not represent a direct modification to the user entity itself.

2. Which permission enum should be used for GET /orgs/{orgid}/apikeys?
I’ve removed the explicit identity.isAdmin check, but it’s unclear which enumeration most accurately represents the intended access policy

For more context, please refer to the discussion above in this PR.

wenyang-cao commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  // === DELETE /v1/orgs/{orgid}/users/{username}/apikeys/{keyid} ===
+  def deleteUserApiKey(identity: Identity, orgid: String, username: String, keyid: String): Route = complete {
```
![image](https://github.com/user-attachments/assets/f7fc8354-b3fc-4830-8112-bcc0034498dc)
Done

wenyang-cao commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -0,0 +1,145 @@
+  ),
+  requestBody = new RequestBody(
+    required = true,
+    content = Array(new Content(
+      mediaType = "application/json",
```
![image](https://github.com/user-attachments/assets/44bd02af-3837-4265-b4d9-ca4931971ec8)
Done

kroczi commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -83,15 +141,81 @@ trait UserApiKeys extends JacksonSupport with AuthenticationSupport {
+                       @Parameter(hidden = true) username: String,
+                       @Parameter(hidden = true) keyid: String): Route = complete {
+
+  db.run(ApiKeysTQ.getById(keyid).result).flatMap {
+    case Nil => 
+        Future.successful(StatusCodes.NotFound)
+    case _ =>
+      db.run(ApiKeysTQ.getById(keyid).delete.asTry).map {
```
do we need to fetch the key twice?

wenyang-cao commented:
Refactor: API Key Route and Test

Behavior and permissions are untouched for now — I notice that in the design doc, hub admin cross-org permissions has been revised, so the expected behavior of several routes should also be changed. However, since the last review mentioned the design of OrgApiKeys might change, I assume that I should hold off on modifying any related logic. Let me know when that’s ready to update.

naphelps commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -0,0 +1,258 @@
+      val rawValue = ApiKeyUtils.generateApiKeyValue()
+      val hashedValue = ApiKeyUtils.sha256Hash(rawValue)
+      val keyId = ApiKeyUtils.generateApiKeyId()
+      val row = ApiKeyRow(organization, keyId, fullId, body.description, hashedValue)
```
Using SHA-256 as a message digest is not sufficient for credential storage. Bcrypt as a legacy technique is more secure than this for this purpose. Current state-of-art is Argon 2 id, or PBKDF2 (FIPS)

cutiechicken replied:
SHA-256 in specifically mentioned in the description in the design doc.  So we may need to update that in the design doc as well.  
For the alternatives you mentioned: 
- bcrypt is already used for password hashing in Password.scala, so using it for API keys would ensure consistency. That said, since the entire auth flow may be refactored in the future, I am not sure whether bcrypt would still be the preferred choice. 
- Argon2id would offer stronger protection, but it requires a native dependency. 
- PBKDF2 is available in the standard library and easier to integrate, but if we use bcrypt for password hashing and PBKDF2 for apikey value hashing it may not be so consistent. 

Let me know which you would prefer please.

wenyang-cao commented on src/main/scala/org/openhorizon/exchangeapi/route/user/User.scala:
```diff
@@ -87,19 +95,35 @@ trait User extends JacksonSupport with AuthenticationSupport {
+
+        db.run(ApiKeysTQ.getByUser(userRow.username).result).map { keys =>
+        val keyMetadata = keys.map(row =>
+          ApiKeyMetadata(
+          id = row.id,
+          description = row.description,
+          owner = null, 
```
In this particular route, the API key list is already under the specific user, so returning the owner (which would just repeat org/username) felt redundant, so I left it as null.

That said, the field is needed in the get-by-id route where the key might be fetched independently. Otherwise, we’d have to define two separate response structs for the get endpoints.

naphelps commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -0,0 +1,145 @@
+      complete {
+        db.run(ApiKeysTQ.insert(row).asTry).map {
+          case Success(_) =>
+            (HttpCode.POST_OK, PostApiKeyResponse(keyId, body.description, username, rawValue))
```
Going to say no. The timestamp is only for the directly connected fields. We see this with Node and its sub-components (status, error, policy..., message, agreement).

naphelps commented on src/main/scala/org/openhorizon/exchangeapi/route/user/User.scala:
```diff
@@ -87,19 +95,35 @@ trait User extends JacksonSupport with AuthenticationSupport {
+      (StatusCodes.OK, GetUsersResponse(usersMap, 0))
+    }
+  } else {
+      Future.successful((StatusCodes.NotFound, GetUsersResponse(Map.empty, 0)))
```
The `Complete` directive handles this, you do not have to wrap this in a Future. its implicit.
